### PR TITLE
Fix Close Connection Deadlock

### PR DESCRIPTION
**Pull request by tonyli508**
Fixing issue #112. 
Thx @vberezkin for the test case.

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@tonyli508 Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/rabbitmq?repositoryId=rabbitmq/rabbitmq-objc-client&pullRequestId=137)!

[Click here](https://cla.pivotal.io/sync/rabbitmq?repositoryId=rabbitmq/rabbitmq-objc-client&pullRequestId=137) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@tonyli508 Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/rabbitmq?repositoryId=rabbitmq/rabbitmq-objc-client&pullRequestId=137)!

**Comment by michaelklishin** [comment]
Thank you. I'm on PTO this week but will try to find a time slot to QA this.

**Comment by michaelklishin** [comment]
I'm having issues running the tests, more specifically I run into a segfault:

> Segmentation fault: 11  xcodebuild test -project RMQClient.xcodeproj -scheme RMQClient -destination 'platform=iOS Simulator,name=iPhone SE,OS=11.3'

will try to upgrade all dependencies there and try cleaning/different platform versions next.

File: RMQClient/RMQConnection.m [hunk 1]
```diff
@@ -321,17 +321,33 @@ - (void)start {
 }
 
 - (void)close {
-    for (RMQOperation operation in self.closeOperations) {
+    [self givingCallingCloseWhileDisconnectedErrorIfNeeded];
+    for (RMQOperation operation in self.safeCloseOperations) {
```
Comment (michaelklishin) [inline RMQClient/RMQConnection.m]: @tonyli508 should we enqueue any operations if we previously invoked an error callback on the delegate here? Other clients usually throw exceptions but this is not how errors are communicated in this case. WDYT?

Comment (tonyli508) [reply RMQClient/RMQConnection.m]: I guess this depends on business decision as well. If you say user need to call close after connection made, then encounter error should just do nothing. If close has higher priority then should just close. In my opinion close has higher priority.

Comment (michaelklishin) [reply RMQClient/RMQConnection.m]: That's a good point. In other words we should try to close the connection because that's the user's intent. I agree.

**Comment by michaelklishin** [comment]
I will merge a separate branch that has some wording and license header changes. Thank you!

**Pull request merged**